### ISSUE TITLE
StopAsync does not wait for tasks completion

### ISSUE THREAD

**Issue by ssougnez** [danielgerlag/workflow-core#645]
Hi,

I created a custom step where I put this:

await Task.Delay(60000);

Then, I start a workflow with this step and I call an API route that does this:

await _wfHost.StopAsync(new CancellationToken());

And this call returns immediately, not waiting for the Task.Delay to finish.
Is it a normal behavior? 
I expected StopAsync to wait for the delay to finish, otherwise, it means that StopAsync is not graceful.

Cheers

**Comment by danielgerlag** [comment]
That sounds like a problem. Will look into it this weekend.

**Comment by danielgerlag** [comment]
Try with v3.2.4 and let me know if that helps?

**Comment by danielgerlag** [comment]
There seems to be something weird with detached child tasks, where the parent task has status `RanToCompletion` before it's really finished.  Try this example.

```c#
        static void  Main(string[] args)
        {
            var task = new Task(async () =>
            {
                
                Console.WriteLine("Hello World!");

                await Task.Delay(3000);

                Console.WriteLine("Done!");
            });
            
            task.Start();
            task.Wait();
            
            Console.WriteLine("END");
            Console.ReadLine();
        }
```

**Comment by ssougnez** [comment]
Hi,

as you can expect, the 3.2.4 did not solve this issue.
I tried your sample and indeed, the behavior seems weird.
On the other hand, I guess there must be some kind of explanation because it would be weird that it's a bug.
I'll try searching Google when I have time to see if there's an explanation to this.

If you find one, I'm interested ;-)

**Comment by ssougnez** [comment]
Just an up comment to notify you that I edited my last one ;-) (will be deleted)

**Comment by danielgerlag** [comment]
Yeah, the problem is that `Task.Run` doesn't have any overloads for passing parameters to the task.

**Comment by danielgerlag** [comment]
@ssougnez give 3.2.5 a spin and let me know?

**Comment by ssougnez** [comment]
Didn't work for me, sorry

**Comment by danielgerlag** [comment]
That's wierd, I tested the example you gave above and it worked. Was that a proxy for the real code?

**Comment by ssougnez** [comment]
Well, the application is pretty big so yeah it's a simplification of the code but basically, I added a task delay in a custom action than call a rest API to shutdown the host.

I did the test in a hurry maybe I missed something. I'll try again tomorrow and tells you what.

**Comment by ssougnez** [comment]
I tested and I confirm that it still does not work for me. I can't give you all the code as it's a full blown application but I'll try to give the main parts.

I have a class `BaseStepAsync` that inherits from `StepBodyAsync`. Basically, it is just to be able to perform some instructions before and after the real instructions of the steps. Indeed, I have to check in the database if a flag is set to `true` and do some logging, so basically, here is the simplified code of this step:

```
namespace Foo
{
    public abstract class BaseStepAsync : StepBodyAsync
    {
        public abstract Task<ExecutionResult> ExecuteAsync(IStepExecutionContext context);

        public override async Task<ExecutionResult> RunAsync(IStepExecutionContext context)
        {
            // Some steps

            var result = await ExecuteAsync(context);

            if (result.Proceed == false)
            {
                return result;
            }

            // Some other steps

            return result;
        }
    }
}
```

Now, custom steps can simply inherit from this class and override `ExecuteAsync`. I modified one of these steps like this:

```
public override async Task<ExecutionResult> ExecuteAsync(IStepExecutionContext context)
{
    await LogInfoAsync("Registering the event");

    await Task.Delay(60000);

    // Content of the action

    return ExecutionResult.Next();
}
```

If I test it, it works, the workflow waits for 1 min before doing the step instructions.
On the same application, I also have a REST API that triggers the following method:

```
public async Task<bool> StartMaintenanceAsync()
{
    var scheduler = await _schedulerFactory.GetScheduler();

    await scheduler.Shutdown(true);

    await _wfHost.StopAsync(new CancellationToken());

    return true;
}
```

Basically, I'm stopping the Quartz scheduler, then I stop the workflow host. This one is just injected in the constructor as `IWorkflowHost`. 

My test is the following: 

* I put a breakpoint at the `Delay` line in the step
* I put a breakpoint in the line that stop the scheduler
* I trigger a workflow
* The first breakpoint is correctly hit
* The next execution of the step does not execute as it is waiting
* I call the REST API
* The second breakpoint is hit
* I press F10 and the execution pointer directly goes on the `return true`, although I'm sure I did that in less than 1 min 

Cheers

**Comment by danielgerlag** [comment]
are you able to encapsulate this into a failing test?  I have not been able to reproduce it.

**Comment by hol353** [comment]
I'm also seeing this behaviour. I'm new to WorkflowCore so might be doing something wrong.

I'm using version 3.3.5. My console app is targetting .Net Core 3.1. My code is below. I expected the code below to write 'Workflow completed' and then 'Program completed'. It only writes 'Program completed' indicating the call to ```host.StopAsync(CancellationToken.None).Wait(); ``` does not wait for the workflow to finish. Suggestions?


```c#
using Microsoft.Extensions.DependencyInjection;
using System;
using System.Threading;
using System.Threading.Tasks;
using WorkflowCore.Interface;
using WorkflowCore.Models;

namespace CloudRunner
{
    class Program
    {
        static void Main(string[] args)
        {
            IServiceProvider serviceProvider = ConfigureServices();

            //start the workflow host
            var host = serviceProvider.GetService<IWorkflowHost>();
            host.RegisterWorkflow<TestWorkflow>();
            host.Start();

            var t = host.StartWorkflow("HelloWorld");

            //Console.Read();

            // Shouldn't this line wait for workflow to complete?
            host.StopAsync(CancellationToken.None).Wait();

            Console.WriteLine("Program completed");
        }

        private static IServiceProvider ConfigureServices()
        {
            // setup dependency injection
            IServiceCollection services = new ServiceCollection();
            services.AddLogging();
            services.AddWorkflow();
            //services.AddWorkflow(x => x.UseMongoDB(@"mongodb://localhost:27017", "workflow"));

            var serviceProvider = services.BuildServiceProvider();

            return serviceProvider;
        }
    }

    public class TestWorkflow : IWorkflow
    {
        public void Build(IWorkflowBuilder<object> builder)
        {
            builder.StartWith(context =>
            {
                Task.Delay(60000);
                return ExecutionResult.Next();
            })
            .Then(context =>
            {
                Console.WriteLine("Workflow completed");
                return ExecutionResult.Next();
            });
        }

        public string Id => "HelloWorld";

        public int Version => 1;
    }
}
```

**Comment by DelineaSamGeorge** [comment]
I'm sorry to necro, but we're having a similar issue to that reported here. If I attempt to gracefully end a workflow host, the host completes its Stop before the underlying Steps have a chance to exit. I don't think I can easily provide an example better than [ssougnez](https://github.com/ssougnez)'s
My specific situation involves running WorkflowCore within a windows service. When the service is stopped, we want to make sure that STOPPED isn't reported before the steps exit. I have tried versions 3.13.0, 3.16.0, and 3.2.6.

**Comment by DelineaSamGeorge** [comment]
This appears to still be occurring on 3.17.0. I am completely prepared for a response that we are using WorkflowCore essentially incorrectly, but I think the behavior I'm seeing is unexpected. I want my step processes to complete or at least attempt to complete when the WorkflowHost is stopped. Instead, that stop seems to try exiting the steps completely without any lifecycle hooks running.

I install and run the following service in Windows. No matter when along the path I stop the service, I would expect MyInnerService to attempt its StopAsync. It does not. I only see the Exec and Start logs. I also tried replacing the hosted MyInnerService with a simple Task.Delay in the step. In that case, too, the Step does not let the delay complete and the workload simply exits out when I stop the service.

Is that expected behavior?

```
Directory.SetCurrentDirectory(SharedPaths.BaseDataDirectory);

var bootstrapBuilder = Host.CreateApplicationBuilder();

var configuration = new ConfigurationBuilder()
    .SetBasePath(AppDomain.CurrentDomain.BaseDirectory)
    .AddJsonFile("appsettings.json", optional: false, reloadOnChange: true).Build();

bootstrapBuilder.Services.AddSingleton<IConfiguration>(configuration);

bootstrapBuilder.Services.AddWorkflow();

bootstrapBuilder.Services
    .AddTransient<MyStep>()
    .AddHostedService<WorkflowHost>();

bootstrapBuilder.Services
    .AddLogging(
        options =>
        {
            var logFile = string.Format("Bootstrap.log");
            var logPath = Path.Combine(SharedPaths.CommonLogDirectory, logFile);
            var logger = new LoggerConfiguration()
                .ReadFrom.Configuration(bootstrapBuilder.Configuration)
                .WriteTo.File(logPath, formatProvider: CultureInfo.InvariantCulture, rollingInterval: RollingInterval.Day, retainedFileCountLimit: 15, shared: true)
                .WriteTo.Console(formatProvider: CultureInfo.InvariantCulture)
                .CreateLogger();
            options
                .ClearProviders()
                .AddSerilog(logger);
        });


bootstrapBuilder.Services
    .AddWindowsService();

var bootstrapper = bootstrapBuilder.Build();

var host = bootstrapper.Services.GetRequiredService<IWorkflowHost>();
host.Options.UseMaxConcurrentWorkflows(1);
host.RegisterWorkflow<MyWorkflow, MyBootstrapState>();

await host.StartWorkflow("Workflow", 1, new MyBootstrapState { });

await bootstrapper.RunAsync();

public class MyWorkflow : IWorkflow<MyBootstrapState>
{
    /// <inheritdoc />
    public string Id => "Workflow";

    /// <inheritdoc />
    public int Version => 1;

    /// <inheritdoc />
    public void Build(IWorkflowBuilder<MyBootstrapState> builder)
    {
        builder
            .While(state => true)
            .Do(recur => recur.Saga(mainSaga =>
            {
                mainSaga
                    .StartWith<MyStep>();
            }));
    }
}

public class MyStep : StepBodyAsync
{
    public override async Task<ExecutionResult> RunAsync(IStepExecutionContext context)
    {
        var bootstrapState = (MyBootstrapState)context.Workflow.Data;

        var webBuilder = WebApplication.CreateBuilder();
        webBuilder.Services
            .AddLogging(configure =>
            {
                var logPath = Path.Combine(SharedPaths.CommonLogDirectory, GetLogFileName());
                var logger = new LoggerConfiguration()
                    .ReadFrom.Configuration(webBuilder.Configuration)
                    .WriteTo.File(logPath, formatProvider: CultureInfo.InvariantCulture, rollingInterval: RollingInterval.Day, retainedFileCountLimit: 15, shared: true)
                    .WriteTo.Console(formatProvider: CultureInfo.InvariantCulture)
                    .CreateLogger();
                configure
                    .ClearProviders()
                    .AddSerilog(logger);
            });

        webBuilder.Services.AddHostedService<MyInnerService>();

        await webBuilder.Build().RunAsync();

        return ExecutionResult.Next();
    }

    public static string GetLogFileName()
    {
        return string.Format("My_Step.log");
    }
}

public class MyInnerService : BackgroundService
{
    public ILogger<MyInnerService> _logger;
    public MyInnerService(ILogger<MyInnerService> logger) { _logger = logger; }

    protected override async Task ExecuteAsync(CancellationToken stoppingToken)
    {
        _logger.LogInformation("STARTING EXEC INNER SERVICE");
        await Task.Delay(5000, stoppingToken);
        _logger.LogInformation("ENDING EXEC INNER SERVICE");
    }

    public override async Task StartAsync(CancellationToken stoppingToken)
    {
        _logger.LogInformation("STARTING START INNER SERVICE");
        await Task.Delay(5000, stoppingToken);
        _logger.LogInformation("ENDING START INNER SERVICE");
    }

    public override async Task StopAsync(CancellationToken stoppingToken)
    {
        _logger.LogInformation("STARTING STOP INNER SERVICE");
        await Task.Delay(5000, stoppingToken);
        _logger.LogInformation("ENDING STOP INNER SERVICE");
    }
}

public class MyBootstrapState
{
}
```
